part-1.py

```diff
@@ -4,17 +4,39 @@
 # the appropriate comment.
 
 # factorial
+def factorial(n):
+    if n < 0:
+        raise ValueError
+    elif n <= 1:
+        return 1
 
+    return n * factorial(n-1)
 
 
 # reverse
-
+def reverse(text):
```

Comment:
👍 Just note that `text[1:]` makes a copy of the original array, element-by-element, so this is O(n^2) in time/space complexity.